support adding custom sites

broofa opened the pull request:
----

Allow users to enter a list of custom domain names that get basic bypass behavior (set referrer, remove cookies), as suggested in #654.

This reworks the options UI into two tabs:
* "Supported Sites" - The existing options UI ("Select All"/"Select None" became checkbox at top of list)
* "Other Sites" - New UI for entering custom domain names, along with a disclaimer about other sites not being supported.

Also includes tweaks for CSS flexbox style for cleaner layout and markup.  See screenshots, below.

----

![image](https://user-images.githubusercontent.com/164050/82837932-75a4ef00-9e7f-11ea-9fb2-a5fd059334d3.png)

![image](https://user-images.githubusercontent.com/164050/82845287-c37a2100-9e98-11ea-8396-42e311ce7050.png)

kimkha commented:
Thanks a lot. I will take a look today.

kimkha commented on src/js/options.js:
```diff
@@ -1,22 +1,54 @@
+
+// Shortcut for document.querySelectorAll()
+function $$(sel, el = document) {
+  return [...el.querySelectorAll(sel)];
```
I wonder what is the oldest version that the extension supported? Check compatible table here: https://caniuse.com/#feat=mdn-javascript_operators_spread_spread_in_arrays

broofa replied:
Switched to `Array.from()`

kimkha replied:
👍

kimkha commented on src/js/options.js:
```diff
@@ -1,22 +1,54 @@
+function selectPane() {
+  const panes = $$('.pane');
+  for (const tab of $$('#tabs button')) {
+    tab.classList.toggle('active', tab == this);
```
`this` is ambiguous, I think you should pass `event` to the function, and retrieve selected tab by `event.target`.

broofa replied:
Switched to use `event.target`

kimkha replied:
👍

iamadamdev commented:
A lot of users have been requesting individual site permissions instead of "<all_urls>", e.g., https://github.com/iamadamdev/bypass-paywalls-chrome/pull/657

If we allow adding custom sites I don't see a way to make it compatible with individual site permissions in the manifest?

kimkha commented:
Why should we restrict site permissions? Any benefit?

sevilla2021 commented:
Firefox add-on also has restricted permissions.

iamadamdev commented:
Is the ability to add custom sites a highly requested feature?

I'm thinking about the following:
The privacy conscious users can have limited site permissions and use the Firefox version without custom sites.
Chrome users can add custom sites and we will leave the site restrictions to <all_urls> as it is now.

iamadamdev commented:
@broofa Can you update this PR so the custom sites also go through the googlebot?

iamadamdev commented:
@broofa Can you make it so the custom sites options only appears for Chrome?

set92 commented:
> 
> 
> Is the ability to add custom sites a highly requested feature?

I think add custom sites is a more highly needed feature than the privacy, because this way anyone can add their own webpages, it comes to mind the regional newspapers or in different languages. For example in my case I use Firefox and in Spain we have few paywalls (like https://www.expansion.com/empresas/banca/2020/06/13/5ee3eb22468aeb740c8b4582.html), so I think it would be better to add this webpage as a custom site instead of asking it to add it globally since most users will not visit this webpages.

Maybe a solution would be 2 versions, one with more privacy setting but without custom sites, and viceversa. Not sure how much work that implies.

iamadamdev commented:
@broofa Thanks for this PR. I've decided I will go ahead and add it for Chrome.
However there are a few more cleanup items to take care of. Almost there.

When dragging extension's folder into Chrome for the first time there is a lot of extra space on the right:
![image](https://user-images.githubusercontent.com/36013816/84602751-74dbf900-ae3e-11ea-9821-122a2eb89397.png)

Please add a small space after the end of the list (after *General Paywall Bypass*), otherwise it almost looks cut off (i.e., the spacing should match the space at the top between "Select all/none" and the first site "Adweek"):
![image](https://user-images.githubusercontent.com/36013816/84602781-9e952000-ae3e-11ea-9fdf-552669fb3219.png)

iamadamdev commented on src/html/options.html:
```diff
@@ -4,35 +4,124 @@
   <style>
+    body {
+      width:350px;
+
```
nit: remove extra line
(Or run semistandard. If not that's okay, I will autoformat this after I merge it)

iamadamdev commented on src/html/options.html:
```diff
@@ -4,35 +4,124 @@
-    uncheck sites you are logged in to otherwise you will be logged out on every visit.
+  <div id="tabs">
+    <button data-pane="sites">Supported Sites</button>
+    <button data-pane="custom">Other Sites</button>
```
Rename to Custom Sites

iamadamdev commented on src/js/options.js:
```diff
@@ -44,27 +76,35 @@ function renderOptions () {
+  for (const el of $$('#tabs button')) {
+    el.addEventListener('click', selectPane);
+  }
+
```
If the browser is not Chrome, remove the Custom Sites tab.
(can use `if (extensionApi !== chrome) { ... }`)

kimkha commented on src/js/background.js:
```diff
@@ -184,9 +184,14 @@ let enabledSites = [];
+  customSites: []
 }, function (items) {
-  enabledSites = Object.values(items.sites);
+  enabledSites = Object.values(items.sites).concat(items.customSites);
+
+  // Use googlebot UA for custom sites
+  useGoogleBotSites = useGoogleBotSites.concat(items.customSites).sort();
```
You should use new variable for `enabledSites` and `useGoogleBotSites`. Otherwise, you can't remove customSites and recover the original list of both variables.

Suggestion could be `_enabledSites` and `_useGoogleBotSites`.

kimkha replied:
You should use new variable for `useGoogleBotSites`. Otherwise, you can't remove customSites and recover the original list of both variables.

Suggestion could be `_useGoogleBotSites`.

iamadamdev commented:
Thanks, confirmed custom sites are not showing up on Firefox.

Can you fix the width on Firefox though?
![image](https://user-images.githubusercontent.com/36013816/84967266-699bfe00-b0c8-11ea-9c8e-f2b659d25b1e.png)

broofa commented:
@iamadamdev Pretty sure the firefox UI width issue is because I changed `
> Thanks, confirmed custom sites are not showing up on Firefox.
> 
> Can you fix the width on Firefox though?

I've made a best-effort attempt at fixing this.  This bug was actually introduced when I fixed https://github.com/iamadamdev/bypass-paywalls-chrome/pull/669#issuecomment-643815201.  It's worth noting that the issue there (width when extension dropped into chrome) was actually not a regression.  It's in `master` as well, but isn't as noticeable because `master` doesn't have borders on any of the DIVs.

Anyhow... take a look, let me know what you think.

iamadamdev commented on src/html/options.html:
```diff
@@ -4,35 +4,134 @@
+  </div>
+
+  <div id="controls">
+    <div id="status"></div>
```
Can you we add some spacing below "Options saved" when it appears?
![image](https://user-images.githubusercontent.com/36013816/85103106-e9a29080-b1ba-11ea-93c2-635a8b371448.png)

broofa replied:
I moved the status DIV to the left of the "Save" button, instead.   This feels cleaner to me, as it avoids causing the dialog to resize.

iamadamdev commented on src/js/options.js:
```diff
@@ -44,27 +76,39 @@ function renderOptions () {
 }
 
+// Select/deselect all supported sites
 function selectAll () {
```
If you click "Select all/none" to select none, then hit save, and reopen, it shows up again as checked.

broofa replied:
Tried to allow for an [indeterminate checkbox state](https://css-tricks.com/indeterminate-checkboxes/), but apparently that [doesn't work](https://bugs.chromium.org/p/chromium/issues/detail?id=1097489) in Chrome extensions. 😢

iamadamdev commented:
Thanks for your contribution! Still need to cleanup some UI now that Chrome style is gone I will merge it for now.
